tariff/pun.go

```diff
@@ -0,0 +1,265 @@
+	"slices"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/cenkalti/backoff/v4"
+	"github.com/evcc-io/evcc/api"
+	"github.com/evcc-io/evcc/util"
+	"golang.org/x/net/html"
+)
+
+type Pun struct {
+	*embed
+	log     *util.Logger
+	charges float32
+	tax     float32
+	data    *util.Monitor[api.Rates]
+}
+
+type NewDataSet struct {
+	XMLName xml.Name `xml:"NewDataSet"`
+	Prezzi  []Prezzo `xml:"Prezzi"`
+}
+
+type Prezzo struct {
+	Data string `xml:"Data"`
+	Ora  string `xml:"Ora"`
+	PUN  string `xml:"PUN"`
+}
+
+type Rate struct {
+	Start time.Time `json:"start"`
+	End   time.Time `json:"end"`
+	Price float64   `json:"price"`
+}
+
+var _ api.Tariff = (*Pun)(nil)
+
+func init() {
+	registry.Add("pun", NewPunFromConfig)
+}
+
+func NewPunFromConfig(other map[string]interface{}) (api.Tariff, error) {
+	cc := struct {
+		Charges float32
+		Tax     float32
+	}{
+		Charges: 0.0,
+		Tax:     0.0,
+	}
+
+	if err := util.DecodeOther(other, &cc); err != nil {
+		return nil, err
+	}
+
+	t := &Pun{
+		log:     util.NewLogger("pun"),
+		charges: cc.Charges,
+		tax:     cc.Tax,
+		data:    util.NewMonitor[api.Rates](2 * time.Hour),
+	}
+
+	done := make(chan error)
+	go t.run(done)
+	err := <-done
+
+	return t, err
+}
+
+func (t *Pun) run(done chan error) {
+	var once sync.Once
+	bo := newBackoff()
+
+	for ; true; <-time.Tick(time.Hour) {
+		var today api.Rates
+		if err := backoff.Retry(func() error {
+			var err error
+
+			today, err = t.getData(time.Now())
+
+			return err
+		}, bo); err != nil {
+			once.Do(func() { done <- err })
+
+			t.log.ERROR.Println(err)
+			continue
+		}
+
+		var tomorrow api.Rates
+		if err := backoff.Retry(func() error {
+			var err error
+
+			tomorrow, err = t.getData(time.Now().AddDate(0, 0, 1))
+
+			return err
+		}, bo); err != nil {
+			once.Do(func() { done <- err })
+			t.log.ERROR.Println(err)
+			continue
+		}
+
+		//merge today and tomorrow data
+		data := append(today, tomorrow...)
+		t.data.Set(data)
+
+		once.Do(func() { close(done) })
+	}
+}
+
+// Rates implements the api.Tariff interface
+func (t *Pun) Rates() (api.Rates, error) {
+	var res api.Rates
+	err := t.data.GetFunc(func(val api.Rates) {
+		res = slices.Clone(val)
+	})
+	return res, err
+}
+
+// Type implements the api.Tariff interface
+func (t *Pun) Type() api.TariffType {
+	return api.TariffTypePriceForecast
+}
+
+func (t *Pun) getData(day time.Time) (api.Rates, error) {
+	// Initial URL
+	urlString := "https://www.mercatoelettrico.org/It/WebServerDataStore/MGP_Prezzi/" + day.Format("20060102") + "MGPPrezzi.xml"
+
+	// Cookie Jar zur Speicherung von Cookies zwischen den Requests
+	jar, _ := cookiejar.New(nil)
+	client := &http.Client{
+		Jar: jar,
+	}
+
+	// Erster Request
+	resp, err := client.Get(urlString)
+	if err != nil {
+		fmt.Println("Error fetching URL:", err)
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	// print resp location
+	fmt.Println(resp.Location())
+
+	body, _ := io.ReadAll(resp.Body)
+	formData, err := parseFormFields(body)
```

Comment:
Igitt- wofür brauchts das?

Reply:
war beim entwickeln für testzwecke

Reply:
ich meinte egtl. das Formulargedöhns?

Reply:
Ja weil man zustimmen muss die Daten nur für private Zwecke zu verwenden. Müssen wir das in EVCC irgendwo angeben auch?

Reply:
Können wir den Reuqestparameter nicht einfach auf "ja" setzen oder muss der wirklich geparsed werden?

Reply:
Ja muss geparst werden, weil da dynamische werte kommen, die man zurücksenden muss.